### ISSUE TITLE
Generate visitors automatically

### ISSUE THREAD

**Issue by jdevera** [dosisod/refurb#35]
I'd like to take stab at this TODO:

```
# TODO: Dynamically generate this somehow instead of hardcoding it
```

in `_visitor_mappings.py`.

Is that okay with you?

**Comment by dosisod** [comment]
Sure! I will note that the visitors are already being automatically generated:

https://github.com/dosisod/refurb/blob/55de2e0617398a96f624f2ba2b964084b7ac9234/refurb/visitor.py#L13-L22

But the generation of the visitors rely on the `MAPPINGS` variable, which is currently hard-coded. I think that is what you meant, just thought I would clarify.

Good luck on implementing this! The way I went about it (which failed) was iterating over the methods in the `TraversalVisitor` class, and pulling out the type info out of there. Curious to see how you go about it!

**Comment by jdevera** [comment]
Yes! I meant generating the `MAPPINGS` variable.

I also tried that approach of iterating over the methods of `TraverseVisitor`. Unfortunately, they all come from the native implementation of mypy and do not contain the typing information for the parameters. I imagine this is what you found and why we don't have them :)

I've been trying some possibly very hacky stuff 👿 , like temporarily changing the module loading mechanism to favour pure python implementations. That worked a bit, but the types came back as strings. I also want to make sure that everything is cleaned up after generation so that the mypy that is used is the fastest available, so that's where I am, seems fun.

**Comment by dosisod** [comment]
Yes, that is exactly what I ran into. I didn't think to change the module loading to load the pure Python implementation, so good job thinking of that! As for the types being returned as strings, how are you getting the types? I (think) the `inspect` module would give you the correct type (as opposed to strings), though I might be mistaken.

**Comment by jdevera** [comment]
The types being strings is the future as of [PEP-563](https://peps.python.org/pep-0563/). However, `typing.get_type_hints` attempts to resolve. There is a whole section on the PEP for the runtime resolution.

What I am finding now is that since I did the pure python loading to get the type data, by the time I've done the cleanup, those type references still refer to the pure python implementation, whereas the rest of the modules are seeing the native implementation, so types are not the same even when they have the same name. Fun times, I swear :)

**Comment by jdevera** [comment]
At the moment `_visitor_mapping.py` is 112 lines and all tests pass. I'll give it a few hackery-reduction passes before the PR :)

**Comment by dosisod** [comment]
Closed via #60
